### PR TITLE
[AArch64] Remove EXT instr before UZP when extracting elements from vector

### PR DESCRIPTION
**Pull request by Lukacma**
Assembly generated  for getting odd/even elements from vector contained extra EXT instruction. This was due to way llvm constructs DAGs when vector_shuffling from larger type to smaller. This patch optimises DAG in these situations, allowing for correct assembly to be emitted.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -21448,6 +21448,29 @@ static SDValue performUzpCombine(SDNode *N, SelectionDAG &DAG,
...
+      Op0.getOperand(0) == Op1.getOperand(0)) {
+
+    SDValue SourceVec = Op0.getOperand(0);
+    uint64_t ExtIdx0 = Op0.getConstantOperandVal(1);
+    uint64_t ExtIdx1 = Op1.getConstantOperandVal(1);
+    uint64_t NumElements = SourceVec.getValueType().getVectorMinNumElements();
+    if (ExtIdx0 == 0 && ExtIdx1 == NumElements / 2) {
+      EVT OpVT = Op0.getOperand(1).getValueType();
+      EVT WidenedResVT = ResVT.getDoubleNumVectorElementsVT(*DAG.getContext());
+      SDValue uzp2 =
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21463]: uzp2 -> Uzp2 (and maybe without the 2? UZP perhaps?)

Can the second SourceVec be Undef if we are only using the bottom half? Or does that produce worse code in places?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -21448,6 +21448,29 @@ static SDValue performUzpCombine(SDNode *N, SelectionDAG &DAG,
...
+          DAG.getNode(N->getOpcode(), DL, WidenedResVT, SourceVec, SourceVec);
+      return DAG.getNode(ISD::EXTRACT_SUBVECTOR, DL, ResVT, uzp2,
+                         DAG.getConstant(0, DL, OpVT));
+    }
+  }
+
+  // following optimization only work with uzp1
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21470]: -> The following optimizations only work with uzp1

**Review by davemgreen — COMMENTED** [review]
Looks like a nice addition.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp (hunk 2 above)
Comment (Lukacma) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:21470]: Done

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Pull request merged**